File: attentive-android-sdk/src/test/java/com/attentive/androidsdk/PersistentStorageTest.java
package com.attentive.androidsdk;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;

import android.content.Context;
import android.content.SharedPreferences;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;

public class PersistentStorageTest {

    private Context context;
    private SharedPreferences sharedPreferences;
    private SharedPreferences.Editor sharedPreferencesEditor;
    private PersistentStorage persistentStorage;

    @Before
    public void setup() {
        sharedPreferences = mock(SharedPreferences.class);
        sharedPreferencesEditor = mock(SharedPreferences.Editor.class);
        doReturn(sharedPreferencesEditor).when(sharedPreferences).edit();

        doReturn(sharedPreferencesEditor).when(sharedPreferencesEditor).putString(any(), any());
        doReturn(sharedPreferencesEditor).when(sharedPreferencesEditor).clear();
        doReturn(sharedPreferencesEditor).when(sharedPreferencesEditor).remove(any());

        context = mock(Context.class);
        doReturn(sharedPreferences).when(context).getSharedPreferences(eq(PersistentStorage.SHARED_PREFERENCES_NAME), anyInt());

        persistentStorage = new PersistentStorage(context);
    }

    @After
    public void checkThatCommitIsNeverCalled() {
        // never call 'commit' because it's not performant since it blocks
        verify(sharedPreferencesEditor, never()).commit();
    }


    @Test
    public void save_validValues_callsEditorPutString() {
        // Arrange
        final String key = "someKey";
        final String value = "someValue";

        // Act
        persistentStorage.save(key, value);

        // Assert
        verify(sharedPreferencesEditor).putString(key, value);
        verify(sharedPreferencesEditor).apply();
    }

    @Test
    public void read_validKey_callsPreferencesGetString() {
        // Arrange
        final String key = "someKey";

        // Act
        persistentStorage.read(key);

        // Assert
        verify(sharedPreferences).getString(key, null);
    }

    @Test
    public void delete_validKey_callsPreferencesEditorRemove() {
        // Arrange
        final String key = "someKey";

        // Act
        persistentStorage.delete(key);

        // Assert
        verify(sharedPreferencesEditor).remove(key);
    }

    @Test
    public void deleteAll_callsPreferencesEditorClear() {
        // Arrange

        // Act
        persistentStorage.deleteAll();

        // Assert
        verify(sharedPreferencesEditor).clear();
    }
}

File: attentive-android-sdk/src/test/java/com/attentive/androidsdk/PersistentStorageTest.kt
package com.attentive.androidsdk

import android.content.Context
import android.content.SharedPreferences
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.mockito.ArgumentMatchers
import org.mockito.Mockito

class PersistentStorageTest {
    lateinit var context: Context
    lateinit var sharedPreferences: SharedPreferences
    lateinit var sharedPreferencesEditor: SharedPreferences.Editor
    lateinit var persistentStorage: PersistentStorage

    @Before
    fun setup() {
        sharedPreferences = Mockito.mock(SharedPreferences::class.java)
        sharedPreferencesEditor = Mockito.mock(SharedPreferences.Editor::class.java)
        Mockito.doReturn(sharedPreferencesEditor).`when`(sharedPreferences)?.edit()

        Mockito.doReturn(sharedPreferencesEditor).`when`(sharedPreferencesEditor)
            ?.putString(ArgumentMatchers.any(), ArgumentMatchers.any())
        Mockito.doReturn(sharedPreferencesEditor).`when`(sharedPreferencesEditor)?.clear()
        Mockito.doReturn(sharedPreferencesEditor).`when`(sharedPreferencesEditor)
            ?.remove(ArgumentMatchers.any())

        context = Mockito.mock(Context::class.java)
        Mockito.doReturn(sharedPreferences).`when`(context)?.getSharedPreferences(
            ArgumentMatchers.eq(PersistentStorage.SHARED_PREFERENCES_NAME),
            ArgumentMatchers.anyInt(),
        )

        persistentStorage = PersistentStorage(context)
    }

    @After
    fun checkThatCommitIsNeverCalled() {
        // never call 'commit' because it's not performant since it blocks
        Mockito.verify(sharedPreferencesEditor, Mockito.never()).commit()
    }

    @Test
    fun save_validValues_callsEditorPutString() {
        // Arrange
        val key = "someKey"
        val value = "someValue"

        // Act
        persistentStorage!!.save(key, value)

        // Assert
        Mockito.verify(sharedPreferencesEditor).putString(key, value)
        Mockito.verify(sharedPreferencesEditor).apply()
    }

    @Test
    fun read_validKey_callsPreferencesGetString() {
        // Arrange
        val key = "someKey"

        // Act
        persistentStorage!!.read(key)

        // Assert
        Mockito.verify(sharedPreferences).getString(key, null)
    }

    @Test
    fun delete_validKey_callsPreferencesEditorRemove() {
        // Arrange
        val key = "someKey"

        // Act
        persistentStorage!!.delete(key)

        // Assert
        Mockito.verify(sharedPreferencesEditor).remove(key)
    }

    @Test
    fun deleteAll_callsPreferencesEditorClear() {
        // Arrange

        // Act

        persistentStorage!!.deleteAll()

        // Assert
        Mockito.verify(sharedPreferencesEditor).clear()
    }
}
